### ISSUE TITLE
Do not close Image in addImageListener method from Camera2ApiManager

### ISSUE THREAD

**Issue by asalmada** [pedroSG94/RootEncoder#1282]
**Describe the bug**

Hi! I'm attempting to use RootEncoder along with MLKit to detect a face while also streaming to a RTMP server. I'm using the `addImageListener` method for RtmpCamera2, but I'm not being able to run [MLKit's process method](https://developers.google.com/ml-kit/vision/face-detection/android#4.-process-the-image) with success. I'm always receiving the error "Internal error has occurred when executing ML Kit tasks".

When searching for this error, [I got this solution](https://stackoverflow.com/a/63406224), informing that we should only call Image's `close` method when MLKit finishes processing it, in a `addOnCompleteListener` callback.

Reading the `addImageListener` method from `Camera2ApiManager.java`, I saw that right after calling the `onImageAvailable` callback with the image as a parameter, the image is closed in the next line. Since Image is not a scalar type (like Int, Double, etc), I believe Java copies a reference of it to the callback, so the image is closed before MLKit can process it.

```
  public void addImageListener(int width, int height, int format, int maxImages, ImageCallback listener) {
    ...
    imageReader.setOnImageAvailableListener(reader -> {
      Image image = reader.acquireLatestImage();
      if (image != null) {
        listener.onImageAvailable(image); // Callback is called here.
        image.close(); // Image is closed right after it. I'd like to close it by myself.
      }
    }, new Handler(imageThread.getLooper()));
    ...
  }
```

Do you think we can have a workaround for it? Thanks!

**Comment by pedroSG94** [comment]
Hello,

I can add a parameter in the method that indicate if it is autoclosed or not but keep in mind that fail doing it (close image) could produce problems like never receive other image.
Maybe the best idea is get data from that image instead of use image itself to work like this:
https://developers.google.com/ml-kit/vision/face-detection/android?hl=en#using-a-bytebuffer-or-bytearray
https://developers.google.com/ml-kit/vision/face-detection/android?hl=en#using-a-bitmap
```java
@Override
      public void onImageAvailable(Image image) {
        ByteBuffer buffer = image.getPlanes()[0].getBuffer();
        byte[] bytes = new byte[buffer.capacity()];
        buffer.get(bytes);
        //you can use bytes or bitmap 
        Bitmap bitmapImage = BitmapFactory.decodeByteArray(bytes, 0, bytes.length, null);
      }
```

This way you can work async mode without worry about close it. Remember that code above must be in the callback in sync mode to avoid problems

**Comment by asalmada** [comment]
Thanks @pedroSG94 for the reply, I'll attempt to apply your suggestion, I also agree that the boolean parameter can be dangerous. I'll let you know by the end of the day if that worked out!

**Comment by asalmada** [comment]
@pedroSG94 it worked, but I also had to change the format, from YUV_420_888 to JPEG. It looks like it's more hard to convert an YUV_420_888 Image, since it contains 3 planes (instead of JPEG that contains only 1).

I chose YUV since the MLKit documentation says "If you use the Camera2 API, capture images in ImageFormat.YUV_420_888 format. If you use the older Camera API, capture images in ImageFormat.NV21 format.", but JPEG is working well for our application.

I know that this is getting out of your repository, but do you think I should attempt to convert YUV to bitmapImage? MLKit docs does not explain why I should stick with YUV.

**Comment by pedroSG94** [comment]
> I chose YUV since the MLKit documentation says "If you use the Camera2 API, capture images in ImageFormat.YUV_420_888 format. If you use the older Camera API, capture images in ImageFormat.NV21 format.", but JPEG is working well for our application.

This is maybe because YUV_420_888 is a RAW type vs JPEG that is compressed so the quality should be better with YUV getting a better accuracy

**Comment by asalmada** [comment]
Hi @pedroSG94,

I'd like to revisit the discussion regarding incorporating a boolean flag to prevent the automatic closure of the Image in `addImageListener`. My recent observations indicate that older devices struggle with the JPEG conversion for each frame when passed as a parameter to the Camera2ApiManager's `addImageListener`. Due to this limitation, I find it necessary to utilize YUV_420 format. Even with an empty image listener, these devices experience significant freezing for extended durations.

Furthermore, I've been receiving warnings from MLKit when using a Bitmap: `"ML Kit has detected that you seem to pass camera frames to the detector as a Bitmap object. This is inefficient. Please use YUV_420_888 format for camera2 API or NV21 format for (legacy) camera API and directly pass down the byte array to ML Kit."`

To summarize, converting an Image from YUV_420 format to a ByteArray or ByteBuffer is not a straightforward task and is extremely processing-intensive.

My proposal is to introduce a boolean flag that controls the automatic closure of the image, with the default value set to true. This addition should not disrupt the current behavior for existing users. For my use case, I will be able to utilize [`InputMedia.fromMediaImage`](https://developers.google.com/android/reference/com/google/mlkit/vision/common/InputImage#fromMediaImage(android.media.Image,%20int)) directly and manage the closure of the Image within MLKit's `addOnCompleteListener` callback.

Looking forward to your thoughts and further discussion on this matter.

**Comment by pedroSG94** [comment]
Hello,

Ok, I will add it but keep in mind that you can avoid MLKit warning using bytearray as suggested in the warning. 

As you can see in my code above you can use bytearray or bitmap removing the last line that create the bitmap and send "byte[] bytes" to MLKit directly.

**Comment by asalmada** [comment]
Hi @pedroSG94,

I've implemented your suggested approach and for that I had to convert each YUV_420_888 Image to NV21 in ByteArray format. This method has proven to be effective. Utilizing Bitmaps was worst due to the additional steps involved, including the need to convert NV21 to JPEG, and then from JPEG to Bitmap.

This solution is performing optimally on older devices, and upon reflection, I agree that introducing a boolean flag may not be necessary in this case.

Thank you for your guidance and support.

**Comment by pedroSG94** [comment]
Hi,

I'm glad that it is working. Anyway I added this method in the last version 2.3.0
